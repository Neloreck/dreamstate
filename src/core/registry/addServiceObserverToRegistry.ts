import { CONTEXT_SERVICES_REGISTRY, CONTEXT_OBSERVERS_REGISTRY } from "@Lib/core/internals";
import { ContextService } from "@Lib/core/management/ContextService";
import { TDreamstateService, TUpdateObserver } from "@Lib/core/types";

import { debug } from "@Macro/debug.macro";

/**
 * Add state changes observer.
 */
export function addServiceObserverToRegistry(
  Service: TDreamstateService,
  observer: TUpdateObserver
): void {
  CONTEXT_OBSERVERS_REGISTRY.get(Service)!.add(observer);

  debug.info("Service observer added:", Service.name);

  // Notify about provision, if it is first observer.
  if (CONTEXT_OBSERVERS_REGISTRY.get(Service)!.size === 1) {
    debug.info("Service provision started:", Service.name);
    (CONTEXT_SERVICES_REGISTRY.get(Service) as ContextService)["onProvisionStarted"]();
  }
}
