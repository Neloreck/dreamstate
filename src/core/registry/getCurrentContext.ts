import { CONTEXT_STATES_REGISTRY } from "@Lib/core/internals";
import { TAnyContextManagerConstructor } from "@Lib/core/types";

import { debug } from "@Macro/debug.macro";

/**
 * Get current supplied context.
 */
export function getCurrentContext<T extends TAnyContextManagerConstructor>(
  Manager: T
): T["prototype"]["context"] | null {
  debug.info("Requested current manager context:", Manager.name);

  return CONTEXT_STATES_REGISTRY.get(Manager) || null;
}
