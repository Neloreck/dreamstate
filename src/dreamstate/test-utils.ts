/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export { ScopeContext as InternalScopeContext } from "@/dreamstate/core/scoping/ScopeContext";

export * from "@/dreamstate/test-utils/registry/getCurrent";
export * from "@/dreamstate/test-utils/utils/getReactConsumer";
export * from "@/dreamstate/test-utils/utils/getReactProvider";

export * from "@/dreamstate/test-utils/registry/mockManagerWithScope";
export * from "@/dreamstate/test-utils/registry/mockScopeProvider";
export * from "@/dreamstate/test-utils/registry/mockScope";

export * from "@/dreamstate/test-utils/utils/nextAsyncQueue";
