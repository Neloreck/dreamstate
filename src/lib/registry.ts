import {
  IContextManagerConstructor,
  TAnyContextManagerConstructor,
  TUpdateObserver,
  TUpdateSubscriber
} from "./types";
import { ContextManager } from "./management";
import {
  CONTEXT_MANAGERS_REGISTRY,
  CONTEXT_OBSERVERS_REGISTRY, CONTEXT_QUERY_METADATA_REGISTRY,
  CONTEXT_SIGNAL_HANDLERS_REGISTRY, CONTEXT_SIGNAL_METADATA_REGISTRY,
  CONTEXT_STATES_REGISTRY,
  CONTEXT_SUBSCRIBERS_REGISTRY,
  IDENTIFIER_KEY,
} from "./internals";
import { onMetadataSignalListenerCalled, subscribeToSignals, unsubscribeFromSignals } from "./signals";

import { log } from "../macroses/log.macro";

/**
 * Create manager ID and init related registry data.
 */
export function createManagerId(description: string): symbol {
  const id: symbol = Symbol(description);

  /**
   * After manager ID resolving all related data should be initialized.
   */
  CONTEXT_STATES_REGISTRY[id as any] = {};
  CONTEXT_OBSERVERS_REGISTRY[id as any] = new Set();
  CONTEXT_SUBSCRIBERS_REGISTRY[id as any] = new Set();
  CONTEXT_SIGNAL_METADATA_REGISTRY[id as any] = [];
  CONTEXT_QUERY_METADATA_REGISTRY[id as any] = [];

  log.info("Context manager id defined:", id, description);

  return id;
}

/**
 * Register context manager entry.
 */
export function registerManager<T extends object>(
  managerConstructor: IContextManagerConstructor<T>,
): void {
  // Only if registry is empty -> create new instance, remember its context and save it to registry.
  if (
    !Object.prototype.hasOwnProperty.call(managerConstructor, IDENTIFIER_KEY) ||
    !Object.prototype.hasOwnProperty.call(CONTEXT_MANAGERS_REGISTRY, managerConstructor[IDENTIFIER_KEY])
  ) {
    const instance: ContextManager<T> = new managerConstructor();

    CONTEXT_STATES_REGISTRY[managerConstructor[IDENTIFIER_KEY]] = instance.context;
    CONTEXT_MANAGERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]] = instance;
    CONTEXT_SIGNAL_HANDLERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]]
      = onMetadataSignalListenerCalled.bind(instance);

    subscribeToSignals(CONTEXT_SIGNAL_HANDLERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]]);

    log.info("Context manager registered:", managerConstructor.name);
  } else {
    log.info("Context manager already registered, continue with old instance:", managerConstructor.name);
  }
}

export function unRegisterManager<T extends object>(
  managerConstructor: IContextManagerConstructor<T>,
): void {
  const instance: ContextManager<T> | undefined = CONTEXT_MANAGERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]];

  if (!instance) {
    log.error("Context manager failed to unregister:", managerConstructor.name);
    throw new Error("Could not find manager instance when unregister it.");
  } else {
    instance["onProvisionEnded"]();

    log.info("Context manager provision ended:", managerConstructor.name);

    // @ts-ignore
    if (!managerConstructor["IS_SINGLE"]) {
      unsubscribeFromSignals(CONTEXT_SIGNAL_HANDLERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]]);

      delete CONTEXT_STATES_REGISTRY[managerConstructor[IDENTIFIER_KEY]];
      delete CONTEXT_MANAGERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]];
      delete CONTEXT_SIGNAL_HANDLERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]];

      log.info("Context manager instance disposed:", managerConstructor.name);
    } else {
      log.info("Context manager instance should not be disposed:", managerConstructor.name);
    }

    log.info("Context manager unregistered:", managerConstructor.name);
  }
}

/**
 * Add state changes observer.
 */
export function addManagerObserverToRegistry<T extends object>(
  managerConstructor: IContextManagerConstructor<T>,
  observer: TUpdateObserver
): void {
  CONTEXT_OBSERVERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]].add(observer);

  log.info("Context manager observer added:", managerConstructor.name);

  // Notify about provision, if it is first observer.
  if (CONTEXT_OBSERVERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]].size === 1) {
    log.info("Context manager provision started:", managerConstructor.name);
    CONTEXT_MANAGERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]]["onProvisionStarted"]();
  }
}

/**
 * Remove state changes observer and kill instance if it is not singleton.
 */
export function removeManagerObserverFromRegistry<T extends object>(
  managerConstructor: IContextManagerConstructor<T>,
  observer: TUpdateObserver
): void {
  // Remove observer.
  CONTEXT_OBSERVERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]].delete(observer);

  log.info("Context manager observer removed:", managerConstructor.name);

  if (CONTEXT_OBSERVERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]].size === 0) {
    unRegisterManager(managerConstructor);
  }
}

/**
 * Subscribe to manager updates/changes.
 * Callback will be fired after each state update.
 */
export function subscribeToManager<T extends object, D extends IContextManagerConstructor<T>>(
  managerConstructor: D,
  subscriber: TUpdateSubscriber<T>,
  loadCurrent: boolean = false
): void {
  CONTEXT_SUBSCRIBERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]].add(subscriber);

  log.info("Context manager subscriber added:", managerConstructor.name, loadCurrent);

  if (loadCurrent) {
    subscriber(CONTEXT_STATES_REGISTRY[managerConstructor[IDENTIFIER_KEY]]);
  }
}

/**
 * Unsubscribe from manager updates.
 */
export function unsubscribeFromManager<T extends object, D extends IContextManagerConstructor<T>>(
  managerConstructor: D,
  subscriber: TUpdateSubscriber<T>
): void {
  log.info("Context manager subscriber removed:", managerConstructor.name);

  CONTEXT_SUBSCRIBERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]].delete(subscriber);
}

/**
 * Get current manager instance from registry.
 * Returns null if nothing is found.
 */
export function getCurrentManager<T extends TAnyContextManagerConstructor>(
  managerConstructor: T
): InstanceType<T> | null {
  log.info("Requested current manager:", managerConstructor.name);

  return CONTEXT_MANAGERS_REGISTRY[managerConstructor[IDENTIFIER_KEY]] as InstanceType<T> || null;
}

/**
 * Get current supplied context.
 */
export function getCurrentContext<S extends object, T extends IContextManagerConstructor<S>>(
  managerConstructor: T
): S | null {
  log.info("Requested current manager context:", managerConstructor.name);

  return CONTEXT_STATES_REGISTRY[managerConstructor[IDENTIFIER_KEY]] as S || null;
}
