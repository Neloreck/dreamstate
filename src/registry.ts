import {
  IContextManagerConstructor,
  IStringIndexed,
  TUpdateObserver,
  TUpdateSubscriber
} from "./types";
import { ContextManager } from "./ContextManager";
import { IDENTIFIER_KEY, SIGNAL_LISTENER_KEY } from "./internals";
import { subscribeToSignals, unsubscribeFromSignals } from "./signals";

export const STORE_REGISTRY: {
  MANAGERS: IStringIndexed<ContextManager<any>>;
  CONTEXT_OBSERVERS: IStringIndexed<Set<TUpdateObserver>>;
  CONTEXT_SUBSCRIBERS: IStringIndexed<Set<TUpdateSubscriber<any>>>;
  CONTEXT_STATES: IStringIndexed<any>;
} = {
  /**
   * Registry of managers instances for global references and better management.
   * Used for current/singleton instances reference.
   */
  MANAGERS: {},
  /**
   * Registry for context observers.
   * Same state across all providers with minimal actions.
   */
  CONTEXT_OBSERVERS: {},
  /**
   * Registry for context subscribers.
   * Subscribers not necessarily related to react
   */
  CONTEXT_SUBSCRIBERS: {},
  /**
   * Registry for current context states for singleton management, beforeUpdate checks and sync state across
   * all manager subscribers.
   */
  CONTEXT_STATES: {},
};

/**
 * Register context manager entry.
 */
export function registerManager<T extends object>(
  managerConstructor: IContextManagerConstructor<T>,
) {
  // Only if registry is empty -> create new instance, remember its context and save it to registry.
  if (
    !Object.prototype.hasOwnProperty.call(managerConstructor, IDENTIFIER_KEY) ||
    !Object.prototype.hasOwnProperty.call(STORE_REGISTRY.MANAGERS, managerConstructor[IDENTIFIER_KEY])
  ) {
    const instance: ContextManager<T> = new managerConstructor();

    STORE_REGISTRY.CONTEXT_STATES[managerConstructor[IDENTIFIER_KEY]] = instance.context;
    STORE_REGISTRY.MANAGERS[managerConstructor[IDENTIFIER_KEY]] = instance;

    subscribeToSignals(instance[SIGNAL_LISTENER_KEY]);
  }
}

export function unRegisterManager<T extends object>(
  managerConstructor: IContextManagerConstructor<T>,
): void {
  const instance: ContextManager<T> | undefined = STORE_REGISTRY.MANAGERS[managerConstructor[IDENTIFIER_KEY]];

  if (!instance) {
    throw new Error("Could not find manager instance when unregister it.");
  } else {
    // @ts-ignore protected.
    instance.onProvisionEnded();
    // @ts-ignore protected field, do not expose it for external usage.
    if (!managerConstructor.IS_SINGLETON) {
      unsubscribeFromSignals(instance[SIGNAL_LISTENER_KEY]);

      delete STORE_REGISTRY.CONTEXT_STATES[managerConstructor[IDENTIFIER_KEY]];
      delete STORE_REGISTRY.MANAGERS[managerConstructor[IDENTIFIER_KEY]];
    }
  }
}

/**
 * Add state changes observer.
 */
export function addManagerObserverToRegistry<T extends object>(
  managerConstructor: IContextManagerConstructor<T>,
  observer: TUpdateObserver
): void {
  // Notify about provision, if it is first observer.
  if (STORE_REGISTRY.CONTEXT_OBSERVERS[managerConstructor[IDENTIFIER_KEY]].size === 0) {
    // @ts-ignore protected and symbol properties.
    STORE_REGISTRY.MANAGERS[managerConstructor[IDENTIFIER_KEY]].onProvisionStarted();
  }

  STORE_REGISTRY.CONTEXT_OBSERVERS[managerConstructor[IDENTIFIER_KEY]].add(observer);
}

/**
 * Remove state changes observer and kill instance if it is not singleton.
 */
export function removeManagerObserverFromRegistry<T extends object>(
  managerConstructor: IContextManagerConstructor<T>,
  observer: TUpdateObserver
): void {
  // Remove observer.
  STORE_REGISTRY.CONTEXT_OBSERVERS[managerConstructor[IDENTIFIER_KEY]].delete(observer);

  if (STORE_REGISTRY.CONTEXT_OBSERVERS[managerConstructor[IDENTIFIER_KEY]].size === 0) {
    unRegisterManager(managerConstructor);
  }
}

export function subscribeToManager<T extends object, D extends IContextManagerConstructor<T>>(
  managerConstructor: D,
  subscriber: TUpdateSubscriber<T>,
  loadCurrent: boolean = false
): void {
  STORE_REGISTRY.CONTEXT_SUBSCRIBERS[managerConstructor[IDENTIFIER_KEY]].add(subscriber);

  if (loadCurrent) {
    subscriber(STORE_REGISTRY.CONTEXT_STATES[managerConstructor[IDENTIFIER_KEY]]);
  }
}

export function unsubscribeFromManager<T extends object, D extends IContextManagerConstructor<T>>(
  managerConstructor: D,
  subscriber: TUpdateSubscriber<T>
): void {
  STORE_REGISTRY.CONTEXT_SUBSCRIBERS[managerConstructor[IDENTIFIER_KEY]].delete(subscriber);
}
