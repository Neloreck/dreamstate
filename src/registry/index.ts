export * from "./addServiceObserverToRegistry";
export * from "./getCurrentContext";
export * from "./getCurrent";
export * from "./registerService";
export * from "./removeServiceObserverFromRegistry";
export * from "./subscribeToManager";
export * from "./unRegisterService";
export * from "./unSubscribeFromManager";
export * from "./getReactContext";
