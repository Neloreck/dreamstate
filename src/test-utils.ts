import { Context } from "react";

import { TAnyContextManagerConstructor, TDreamstateWorker, TUpdateObserver } from "./types";
import { ContextWorker } from "./management";
import {
  addWorkerObserverToRegistry,
  registerWorker as internalRegisterWorker,
  removeWorkerObserverFromRegistry,
  unRegisterWorker as internalUnRegisterWorker
} from "./registry";
import { CONTEXT_OBSERVERS_REGISTRY, CONTEXT_WORKERS_REGISTRY } from "./internals";

/**
 * Register worker class.
 */
export function registerWorker<T extends TDreamstateWorker>(workerClass: T): InstanceType<T> {
  if (!workerClass || !workerClass.prototype || !(workerClass.prototype instanceof ContextWorker)) {
    throw new TypeError("Cannot register invalid worker. Expected class extending ContextWorker.");
  }

  internalRegisterWorker(workerClass);

  return CONTEXT_WORKERS_REGISTRY.get(workerClass) as InstanceType<T>;
}

/**
 * Unregister worker.
 */
export function unRegisterWorker<T extends TDreamstateWorker>(
  workerClass: T,
  forceUnregister: boolean = false
): void {
  if (!workerClass || !workerClass.prototype || !(workerClass.prototype instanceof ContextWorker)) {
    throw new TypeError("Cannot register invalid worker. Expected class extending ContextWorker.");
  }

  internalUnRegisterWorker(workerClass, forceUnregister);
}

/**
 * Get current worker observerrs count or null if it is not registered.
 * Observers - all points that create react context provision.
 */
export function getWorkerObserversCount<T extends TDreamstateWorker>(
  workerClass: T
): number | null {
  if (CONTEXT_OBSERVERS_REGISTRY.has(workerClass)) {
    return CONTEXT_OBSERVERS_REGISTRY.get(workerClass)!.size;
  } else {
    return null;
  }
}

/**
 * Check if current worker is provided.
 */
export function isWorkerProvided<T extends TDreamstateWorker>(
  workerClass: T
): boolean {
  return Boolean(getWorkerObserversCount(workerClass));
}

/**
 * Add context manager observer and trigger all related events (onProvisionStarted for first observer).
 */
export function addManagerObserver<T extends TAnyContextManagerConstructor>(
  managerConstructor: T,
  observer: TUpdateObserver
) {
  addWorkerObserverToRegistry(managerConstructor, observer);
}

/**
 * Add context manager observer and trigger all related events (onProvisionEnded for last observer).
 */
export function removeManagerObserver<T extends TAnyContextManagerConstructor>(
  managerConstructor: T,
  observer: TUpdateObserver
) {
  removeWorkerObserverFromRegistry(managerConstructor, observer);
}

/**
 * Get react provider of selected context manager.
 */
export function getReactConsumer<T extends TAnyContextManagerConstructor>(
  managerConstructor: T
): Context<T["prototype"]["context"]>["Consumer"] {
  return managerConstructor.REACT_CONTEXT.Consumer;
}

/**
 * Get react provider of selected context manager.
 */
export function getReactProvider<T extends TAnyContextManagerConstructor>(
  managerConstructor: T
): Context<T["prototype"]["context"]>["Provider"] {
  return managerConstructor.REACT_CONTEXT.Provider;
}

/**
 * Wait for all queued events.
 * Promisified setTimeout(,0).
 */
export function nextAsyncQueue(): Promise<void> {
  return new Promise((resolve: () => void) => setTimeout(resolve));
}
