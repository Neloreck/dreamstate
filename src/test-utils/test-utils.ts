/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export * from "@Lib/test-utils/utils/getReactConsumer";
export * from "@Lib/test-utils/utils/getReactProvider";
export * from "@Lib/test-utils/utils/nextAsyncQueue";

export * from "@Lib/test-utils/registry/addServiceObserver";
export * from "@Lib/test-utils/registry/getServiceObserversCount";
export * from "@Lib/test-utils/registry/isServiceProvided";
export * from "@Lib/test-utils/registry/registerService";
export * from "@Lib/test-utils/registry/removeServiceObserver";
export * from "@Lib/test-utils/registry/unRegisterService";
