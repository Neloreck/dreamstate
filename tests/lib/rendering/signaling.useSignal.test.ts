import { createElement } from "react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";

import { getCurrentManager } from "@Lib/registry";

import { nextAsyncQueue, registerManagerClass, unRegisterManagerClass } from "@Tests/helpers";
import { EmittingContextManager, ESignal, TStringSignalEvent, UsingSignalFunction } from "@Tests/assets";

describe("Signals and signaling.", () => {
  beforeEach(() => {
    registerManagerClass(EmittingContextManager);
  });

  afterEach(() => {
    unRegisterManagerClass(EmittingContextManager);
  });

  it("Functional components should properly subscribe to signals and update ui.", async () => {
    const emittingContextManager: EmittingContextManager = getCurrentManager(EmittingContextManager)!;

    const mockFn = jest.fn((signalEvent: TStringSignalEvent) => {
      expect(signalEvent.type).toBe(ESignal.STRING_SIGNAL);
      expect(signalEvent.emitter).toBe(emittingContextManager);
      expect(signalEvent.data).toBe("newValue");
    });
    const tree = mount(createElement(UsingSignalFunction, { onInternalSignal: mockFn }));

    expect(tree).toMatchSnapshot();

    await act(async () => {
      emittingContextManager.sendStringSignal("newValue");

      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(mockFn).toHaveBeenCalled();

    mockFn.mockClear();

    tree.unmount();

    emittingContextManager.sendStringSignal("newValue");
    await nextAsyncQueue();

    expect(mockFn).not.toHaveBeenCalled();
  });
});
