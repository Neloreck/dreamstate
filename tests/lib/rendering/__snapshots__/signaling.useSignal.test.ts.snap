// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signals and signaling. Functional components should properly subscribe to signals and update ui. 1`] = `
<UsingSignalFunction
  onInternalSignal={[MockFunction]}
>
  <div>
    initial
  </div>
</UsingSignalFunction>
`;

exports[`Signals and signaling. Functional components should properly subscribe to signals and update ui. 2`] = `
<UsingSignalFunction
  onInternalSignal={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "cancel": [Function],
            "data": "newValue",
            "emitter": EmittingContextManager {
              "context": Object {},
            },
            "type": "STRING_SIGNAL",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div>
    newValue
  </div>
</UsingSignalFunction>
`;
